src/server/pkg/storage/fileset/merge.go

```diff
@@ -28,153 +27,169 @@ func newMergeReader(chunks *chunk.Storage, fileSets []FileSet) *MergeReader {
-		// Collect files from file streams.
-		var files []File
+		var fss []*fileStream
 		for _, s := range ss {
-			fs := s.(*fileStream)
-			files = append(files, fs.file)
+			fss = append(fss, s.(*fileStream))
+		}
+		if skipPrefix != "" && strings.HasPrefix(fss[0].key(), skipPrefix) {
+			for i, fs := range fss {
+				if fs.streamPriority() < skipPriority {
+					fss = fss[:i]
+					break
+				}
+			}
 		}
-		if len(files) == 1 {
-			return cb(files[0])
+		if len(fss) == 0 {
+			return nil
+		}
+		if IsDir(fss[0].key()) {
+			skipPrefix = fss[0].key()
+			skipPriority = fss[0].streamPriority()
+			return nil
+		}
+		if len(fss) == 1 {
+			if fss[0].deletive {
+				return nil
+			}
+			return cb(newFileReader(ctx, mr.chunks, fss[0].file.Index()))
+		}
+		idx := mergeFile(fss)
+		// Handle a full delete.
+		if len(idx.File.Parts) == 0 {
+			return nil
 		}
-		idx := mergeFiles(files)
 		return cb(newMergeFileReader(ctx, mr.chunks, idx))
 	})
 }
 
-type fileStream struct {
-	iterator *Iterator
-	file     File
-	priority int
-}
-
-func (fs *fileStream) next() error {
-	var err error
-	fs.file, err = fs.iterator.Next()
-	return err
-}
-
-func (fs *fileStream) key() string {
-	return fs.file.Index().Path
-}
-
-func (fs *fileStream) streamPriority() int {
-	return fs.priority
-}
-
-func mergeFiles(fs []File) *index.Index {
+func mergeFile(fss []*fileStream) *index.Index {
 	mergeIdx := &index.Index{
-		FileOp: &index.FileOp{},
-	}
-	mergeIdx.Path = fs[0].Index().Path
-	// Handle a full delete.
-	if fs[0].Index().FileOp.Op == index.Op_DELETE {
-		mergeIdx.FileOp.Op = index.Op_DELETE
-		return mergeIdx
+		Path: fss[0].file.Index().Path,
+		File: &index.File{},
 	}
-	// Collect the indexes for the files being merged.
-	var idxs []*index.Index
-	for _, f := range fs {
-		idx := f.Index()
-		if idx.FileOp.Op == index.Op_DELETE {
-			mergeIdx.FileOp.Op = index.Op_OVERWRITE
+	var ps []*partStream
+	for _, fs := range fss {
+		idx := fs.file.Index()
+		if fs.deletive && idx.File.Parts == nil {
 			break
 		}
-		idxs = append(idxs, idx)
-		if idx.FileOp.Op == index.Op_OVERWRITE {
-			mergeIdx.FileOp.Op = index.Op_OVERWRITE
-			break
-		}
-	}
-	// Collect the header data op from the highest priority file.
-	hdo := getHeaderDataOp(idxs[0].FileOp.DataOps)
-	// Collect the content data ops, then merge based on the lexicograhical ordering of the tags.
-	var dos [][]*index.DataOp
-	for _, idx := range idxs {
-		dos = append(dos, getContentDataOps(idx.FileOp.DataOps))
+		ps = append(ps, &partStream{
+			parts:    idx.File.Parts,
+			deletive: fs.deletive,
+		})
 		mergeIdx.SizeBytes += idx.SizeBytes
 	}
-	mergeIdx.FileOp.DataOps = append([]*index.DataOp{hdo}, mergeDataOps(dos)...)
+	// Merge the parts based on the lexicograhical ordering of the tags.
+	mergeIdx.File.Parts = mergeParts(ps)
 	return mergeIdx
 }
 
-func mergeDataOps(dataOps [][]*index.DataOp) []*index.DataOp {
-	if len(dataOps) == 0 {
+func mergeParts(pss []*partStream) []*index.Part {
```

Comment:
It looks like we are performing a merge on streams which are entirely in memory.  Essentially performing heap sort on a slice.  Could we use the "sort" package instead? Is there a path to having these stream from out of memory that we are preparing for?

Reply:
The priority queue implementation handles the heap sort and captures the behavior of collecting entries with the same keys and ordering them by priority (which is important for ordering additive / deletive entries). Although the streaming functionality is not being taken advantage of, I think the features I mentioned above and the simple interface (although I think there is opportunity for improvement here) make it a good choice here.